src/change_res.cpp

```diff
@@ -56,3 +68,193 @@ void ChangeRes(int, int, int, int)
+                int scale_factor = int_h / 600;
+                int_h /= scale_factor;
+            }
+        }
+
+        // minimum height constraint
+        if(int_h < 320)
+            int_h = 320;
+
+        // maximum height constraint
+        if(int_h > 720 && req_h <= 720)
+            int_h = 720;
+
+        // now, set width based on height and scaling mode
+        if(g_videoSettings.scaleMode == SCALE_FIXED_05X)
+        {
+            int_w = int_w * 2;
+        }
+        else if(g_videoSettings.scaleMode == SCALE_FIXED_1X)
+        {
+            int_w = int_w;
+        }
+        else if(g_videoSettings.scaleMode == SCALE_FIXED_2X)
+        {
+            int_w = int_w / 2;
+        }
+        else if(g_videoSettings.scaleMode == SCALE_DYNAMIC_INTEGER)
+        {
+            int scale_factor = orig_int_h / int_h;
+            if(scale_factor == 0)
+            {
+                int_w = int_w;
+            }
+            else if(int_w / scale_factor >= 800)
+            {
+                int_w = int_w / scale_factor;
+            }
+            else
+            {
+                // scale based on width
+                int scale_factor = int_w / 800;
+                if(scale_factor != 0)
+                    int_w = int_w / scale_factor;
+
+                // rescale the height if possible
+                if(scale_factor != 0 && req_h == 0)
+                {
+                    int_h = orig_int_h / scale_factor;
+                    if(int_h < 600)
+                        int_h = 600;
+                    if(int_h > 720)
+                        int_h = 720;
+                }
+            }
+        }
+        else
+        {
+            int_w = (int_w * int_h) / orig_int_h;
+        }
+
+        // force >800x600 resolution if required
+        if((!g_compatibility.free_level_res && !LevelSelect && !GameMenu)
+            || (!g_compatibility.free_world_res && LevelSelect && !GameMenu))
+        {
+            if(int_w < 800)
+            {
+                int_h = int_h * 800 / int_w;
+                int_w = 800;
+                if(int_h > 720)
+                    int_h = 720;
+            }
+            if(int_h < 600)
+            {
+                int_w = int_w * 600 / int_h;
+                int_h = 600;
+            }
+        }
+
+        // minimum width constraint
+        if(int_w < 480)
+            int_w = 480;
+
+        // maximum 2.4 (cinematic) aspect ratio
+        if(int_w > int_h*2.4)
+            int_w = (int)(int_h*2.4);
+
+        // force even dimensions
+        int_w -= int_w & 1;
+        int_h -= int_h & 1;
+
+        ScreenW = int_w;
+        ScreenH = int_h;
+    }
+    else
+    {
+        ScreenW = req_w;
+        ScreenH = req_h;
+    }
+#endif // #ifndef THEXTECH_FIXED_RES
+
+    XRender::updateViewport();
+
+    // recenter the game menu graphics
+    if(GameMenu)
+    {
+        SetupScreens();
+        GameMenu = false;
+        GetvScreenAverage();
+        GameMenu = true;
```

Comment:
I didn't want to add the argument since it is almost always the correct behavior to use `GameMenu` and looking it up as a global before the function call and passing it as an argument is more expensive than directly looking up the global in the function. But we can also change it to an argument if you think this is very bad style, or I can add a comment indicating why this was done this way.